### PR TITLE
Fcitx5: add fcitx5-rime, and upgrade librime

### PR DESCRIPTION
**Pull request by bekcpear**
1. Upgrade fcitx-rime to version 5 ~~, it conflicts with fcitx-rime:4 due to the icons~~.

2. ~~Also add a new slot 'essay' for rime-data with a newer version.~~

    ~~The previous repo 'brise' for rime-data is deprecated as of Apr, 2018. However, the new rime-data is spread over several repositories, and most of them does not have a release version (including essential data). So I have to use a different slot to make it possible to select this rime-data which has a fake version number (0.0.0.[date]) that satisfies Semantic Versioning.~~
    ~~The package name prefix is suggested by the upstream: https://github.com/rime/plum#install-as-shared-data~~

    <details>
    <summary><s>tree:</s></summary>

    ```
    app-i18n/fcitx-rime:5
        \--- app-i18n/rime-data:essay
              |--- app-i18n/rime-data-prelude (essential)
              |--- app-i18n/rime-data-array
              |--- app-i18n/rime-data-bopomofo
              |--- app-i18n/rime-data-cangjie
              |--- app-i18n/rime-data-combo-pinyin
              |--- app-i18n/rime-data-double-pinyin
              |--- app-i18n/rime-data-emoji
              |--- app-i18n/rime-data-ipa
              |--- app-i18n/rime-data-luna-pinyin
              |--- app-i18n/rime-data-middle-chinese
              |--- app-i18n/rime-data-pinyin-simp
              |--- app-i18n/rime-data-quick
              |--- app-i18n/rime-data-scj
              |--- app-i18n/rime-data-soutzoe
              |--- app-i18n/rime-data-stenotype
              |--- app-i18n/rime-data-stroke
              |--- app-i18n/rime-data-terra-pinyin
              |--- app-i18n/rime-data-wubi
              |--- app-i18n/rime-data-wugniu
    ```

    </details>

    ~~The [rime-cantonese](https://github.com/rime/rime-cantonese) input schema is not added here duo to the ODbL license is not included in gentoo yet, this input schema may be added in the future.~~

3. upgrade app-i18n/librime to 1.9.0
    This upgrade skipped over a minor version, and the time span was too long, so only the ~amd64 and ~x86 keywords are kept.

### CODE REVIEW

File: app-i18n/rime-data-array/rime-data-array-0.0.0.20221120.ebuild [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+RIME_DATA_DIR="/usr/share/rime-data"
+DOCS=(
+	AUTHORS
+	LICENSE
+	README.md
+)
+
+src_install() {
+	insinto "$RIME_DATA_DIR"
```
Comment (liuyujielol) [inline app-i18n/rime-data-array/rime-data-array-0.0.0.20221120.ebuild:30]: As per https://projects.gentoo.org/qa/policy-guide/ebuild-format.html#pg0101
Please change other ebuilds too.
```suggestion
	insinto "${RIME_DATA_DIR}"
```

File: app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+DEPEND="
+	!app-i18n/fcitx-rime:4
+	>=app-i18n/fcitx-5.1.2:5
+	app-i18n/librime
+	app-i18n/rime-data:essay
+	virtual/libintl
+"
+RDEPEND="${DEPEND}"
+BDEPEND="
+	kde-frameworks/extra-cmake-modules
```
Comment (liuyujielol) [inline app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild:26]: The slotmove ECM:5 to ECM:0  is for upcoming KDE6, maybe we should depend ECM with slot.

File: app-i18n/librime/librime-1.9.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+src_configure() {
+	local -x CXXFLAGS="${CXXFLAGS} -I${ESYSROOT}/usr/include/utf8cpp"
+
+	# for glog
+	if use debug; then
+		CXXFLAGS+=" -DDCHECK_ALWAYS_ON"
+	else
+		CXXFLAGS+=" -DNDEBUG"
+	fi
```
Comment (liuyujielol) [inline app-i18n/librime/librime-1.9.0.ebuild:43]: Can't `CMAKE_BUILD_TYPE` do the job?

File: app-i18n/librime/librime-1.9.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	dev-cpp/yaml-cpp:0=
+	dev-libs/leveldb:0=
+	dev-libs/marisa:0=
+"
+DEPEND="${RDEPEND}
+	dev-libs/darts
+	dev-libs/utfcpp
+	test? ( dev-cpp/gtest )
+"
+BDEPEND=""
```
Comment (liuyujielol) [inline app-i18n/librime/librime-1.9.0.ebuild:31]: EmptyGlobalAssignment

File: app-i18n/librime/librime-1.9.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+DOCS=(CHANGELOG.md README.md)
+
+src_configure() {
+	local -x CXXFLAGS="${CXXFLAGS} -I${ESYSROOT}/usr/include/utf8cpp"
```
Comment (liuyujielol) [inline app-i18n/librime/librime-1.9.0.ebuild:36]: Can we address what this is for?

File: app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	stenotype
+	+stroke
+	+terra-pinyin
+	wubi
+	wugniu
+"
+
+DEPEND="!app-i18n/rime-data:0"
+RDEPEND="${DEPEND}"
+PDEPEND="
```
Comment (liuyujielol) [inline app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild:39]: Why these are `PDEPEND`? Look like `RDEPEND` to me.

File: app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+MY_PN="rime-essay"
+MY_COMMIT="e0519d0579722a0871efb68189272cba61a7350b"
+
+DESCRIPTION="Data resources for Rime Input Method Engine"
+HOMEPAGE="https://rime.im/ https://github.com/rime/plum"
+SRC_URI="https://github.com/rime/${MY_PN}/archive/${MY_COMMIT}.tar.gz -> ${MY_PN}-${PV}.tar.gz"
+
+LICENSE="LGPL-3"
+SLOT="essay"
+KEYWORDS="~amd64 ~arm64 ~loong ~ppc ~ppc64 ~riscv ~x86"
```
Comment (liuyujielol) [inline app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild:15]: Maybe we should start with `~amd64 ~x86`

File: app-i18n/librime/librime-1.9.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		CXXFLAGS+=" -DDCHECK_ALWAYS_ON"
+	else
+		CXXFLAGS+=" -DNDEBUG"
+	fi
+
+	local mycmakeargs=(
+		-DBUILD_TEST=$(usex test ON OFF)
+		-DENABLE_EXTERNAL_PLUGINS=ON
+		-DINSTALL_PRIVATE_HEADERS=ON
+	)
```
Comment (liuyujielol) [inline app-i18n/librime/librime-1.9.0.ebuild:49]: Does not respect `MAKEOPTS`
https://github.com/rime/librime/blob/a60876745af20ecc8489ec6997c6c195949b99ac/CMakeLists.txt#L162-L168
Please `inherit multiprocessing`

```suggestion
		-DCMAKE_BUILD_PARALLEL_LEVEL=$(makeopts_jobs)
	)
```

File: app-i18n/rime-data-array/rime-data-array-0.0.0.20221120.ebuild (hunk 1 above)
Comment (bekcpear) [reply app-i18n/rime-data-array/rime-data-array-0.0.0.20221120.ebuild:30]: In fact, there is no difference between them in the case of the use of a single variable, does the QA reports some error?

Comment (liuyujielol) [reply app-i18n/rime-data-array/rime-data-array-0.0.0.20221120.ebuild:30]: Quotes from PG0101
`Rationale: the recommended constructs are less error-prone. Consistency avoids unnecessary changes when other 
developers edit the ebuild.`

File: app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild (hunk 2 above)
Comment (bekcpear) [reply app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild:26]: There is only one slot for this package, why do we need to specify a default slot?

File: app-i18n/rime-data-array/rime-data-array-0.0.0.20221120.ebuild (hunk 1 above)
Comment (bekcpear) [reply app-i18n/rime-data-array/rime-data-array-0.0.0.20221120.ebuild:30]: Okay

File: app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild (hunk 2 above)
Comment (liuyujielol) [reply app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild:26]: As mentioned, it's for upcoming KDE6
https://github.com/gentoo/gentoo/commit/4396c608794f2d2760ef53df83426c52aef59704

Comment (bekcpear) [reply app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild:26]: But now that there's only one solt, and there's no evidence that additional slots will be added later, is there anything wrong or improper with letting ebuild choose the default?

Comment (liuyujielol) [reply app-i18n/fcitx-rime/fcitx-rime-5.1.3.ebuild:26]: Guess I got it wrong, forget about this.

File: app-i18n/librime/librime-1.9.0.ebuild (hunk 3 above)
Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:43]: Not sure if CMAKE_BUILD_TYPE can achieve the same situation. These are defined in the glog library.

File: app-i18n/librime/librime-1.9.0.ebuild (hunk 5 above)
Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:36]: This is for the utfcpp library which is not included in 1.7.3, however with 1.9.0, the utfcpp and darts headers are included in the package it self, so I will remove it and above DEPEND, thanks for this good debug!

File: app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild (hunk 6 above)
Comment (bekcpear) [reply app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild:39]: Please refer to https://devmanual.gentoo.org/general-concepts/dependencies/#post-dependencies

All these data packages except essay itself and prelude are not necessary for the running of rime. So I will put prelude into RDEPEND and keep others unchanged. Thanks!

File: app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild (hunk 7 above)
Comment (bekcpear) [reply app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild:15]: These are all architecture independent packages (data only), so I decide to add more keywords to satisfy ibus-rime and possible fcitx-rime:4. I think it's fine.

File: app-i18n/librime/librime-1.9.0.ebuild (hunk 8 above)
Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:49]: Yes, it's my big mistake, thank you very much!

File: app-i18n/librime/librime-1.9.0.ebuild (hunk 3 above)
Comment (liuyujielol) [reply app-i18n/librime/librime-1.9.0.ebuild:43]: `-DDCHECK_ALWAYS_ON` is for glog right? But the librime is linked against system `dev-cpp/glog`.
Upstream git submodule is not include in src tarball. Is setting this flag do anything useful?

Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:43]: If I understand correctly, it includes Glog's header file (logging.h) at compile time, uses the DLOG macro, so it's valid.

File: app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild (hunk 6 above)
Comment (liuyujielol) [reply app-i18n/rime-data/rime-data-0.0.0.20230204.ebuild:39]: Can we make it build against system headers?
That is `dev-libs/darts` `dev-libs/utfcpp` `x11-libs/xcb-util-keysyms`.
If not, we should include LICENSE for corresponding headers.

File: app-i18n/librime/librime-1.9.0.ebuild (hunk 5 above)
Comment (liuyujielol) [reply app-i18n/librime/librime-1.9.0.ebuild:36]: Can we make it build against system headers?
That is `dev-libs/darts` `dev-libs/utfcpp` `x11-libs/xcb-util-keysyms`.
If not, we should include LICENSE for corresponding headers.

Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:36]: I tend to use the included header files, also I will append additional licenses.

File: app-i18n/librime/librime-1.9.0.ebuild (hunk 3 above)
Comment (liuyujielol) [reply app-i18n/librime/librime-1.9.0.ebuild:43]: Still, I think `CMAKE_BUILD_TYPE=debug` should be set for `debug` & append-cxxflags, and do nothing for normal build.

Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:40]: Fixed, Thanks!

Comment (liuyujielol) [reply app-i18n/librime/librime-1.9.0.ebuild:40]: https://github.com/gentoo/gentoo/blob/ede01cc9457106602b6f48aa9e76d06c07dd512e/eclass/cmake.eclass#L47-L58
`CMAKE_BUILD_TYPE` is set in cmake.eclass, `CMAKE_BUILD_TYPE=Gentoo` is only for EAPI=7
Something like `use debug && CMAKE_BUILD_TYPE=Debug` is better.

Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:41]: You are right, fixed.

File: app-i18n/librime/librime-1.9.0.ebuild [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	dev-libs/marisa:0=
+"
+DEPEND="${RDEPEND}
+	test? ( dev-cpp/gtest )
+"
+
+DOCS=(CHANGELOG.md README.md)
+
+src_configure() {
+	local -x CXXFLAGS="${CXXFLAGS}"
```
Comment (liuyujielol) [inline app-i18n/librime/librime-1.9.0.ebuild:33]: Any reason to keep this?
If it's for `CXXFLAGS+=" "`,  use `append-flags` or `append-cxxflags` from flag-o-matic.eclass

Comment (bekcpear) [reply app-i18n/librime/librime-1.9.0.ebuild:33]: Just keep it simple, there is no good reason to inherit an extra eclass for a temporary variable.

**Comment by bekcpear** [comment]
Rebased  to the updated master branch due to the file conflicting with `ibus-rime-1.5.0.ebuild`.

**Comment by bekcpear** [comment]
Upgrade fcitx-rime:5 to 5.1.4

Due to the icon files renamed, removing the conflicting with fcitx-rime:4.

**Comment by bekcpear** [comment]
Give up upgrading the rime-data.

**Pull request closed without merging**